sdks/python/apache_beam/io/gcp/bigquery.py

```diff
@@ -1056,23 +1054,43 @@ def deserialize_rows(self, read_rows_response, reader):
 
   def read(self, range_tracker):
     storage_client = bq_storage.BigQueryReadClient()
-    read_rows_iterator = iter(storage_client.read_rows(self.read_stream_name))
-    read_rows_response = next(read_rows_iterator)
-    avro_schema = avro.schema.Parse(read_rows_response.avro_schema.schema)
-    reader = avroio.DatumReader(avro_schema)
-    while True:
-      # self.deserialize_rows(read_rows_response, reader)
-      decoder = avroio.BinaryDecoder(
-          io.BytesIO(read_rows_response.avro_rows.serialized_binary_rows))
-      current_row = 0
-      while current_row < read_rows_response.row_count:
-        self.deserialized_rows.append(reader.read(decoder))
-        current_row += 1
-      read_rows_response = next(read_rows_iterator, None)
-      if read_rows_response is None:
-        break
-
-    return iter(self.deserialized_rows)
+    row_reader = _ReadRowsResponseReader(
+        storage_client.read_rows(self.read_stream_name))
+    return iter(row_reader)
+
+
+class _ReadRowsResponseReader():
+  def __init__(self, read_rows_responses):
+    self.read_rows_iterator = iter(read_rows_responses)
+    self.read_rows_response = next(self.read_rows_iterator)
```

Comment:
Nit: it's possible in the case of very selective filters that you might get no responses when reading from a stream. You might consider deferring some of this initialization to the first `__next__` call.

Reply:
There doesn't seem to be a simple of handling this case in the `__next__` call. I've gone ahead and added some error handling to account for the possibility of an empty response. Added a test for the same.